File: python/lib/keyvalue.py
"""
data tools
"""


class KeyValue:

    def __init__(self, key, value, delimitter=None, keystring=None):
        self.key        = key
        self.value      = value
        if keystring is None:
            if key is None:
                self.keystring = None
            else:
                if delimitter is None:
                    self.keystring = key + ' '
                else:
                    self.keystring = key + ' ' + delimitter + ' '
        else:
            self.keystring = keystring

    def valid(self):
        return self.key is not None

    def has_var(self):
        return "${" in self.value

    @staticmethod
    def fromAssign(linestr, delimitter='='):
        keyval = linestr.split(delimitter, 1)
        if len(keyval) > 1:
            key   = keyval[0].strip()
            value = keyval[1].strip()
            return KeyValue(key, value, delimitter)
        else:
            return KeyValue(None, None)

    @staticmethod
    def fromAssignString(linestr, delimitter='='):
        keyval = linestr.split(delimitter, 1)
        if len(keyval) > 1:
            key    = keyval[0].strip()
            text   = keyval[1].strip().split('"', 2)
            value  = text[1]
            return KeyValue(key, value, delimitter)
        else:
            return KeyValue(None, None)

    @staticmethod
    def fromAssignCharArray(linestr, delimitter='='):
        keyval = linestr.split(delimitter, 1)
        if len(keyval) > 1:
            key    = keyval[0].strip()
            text   = keyval[1].strip().split("'", 2)
            value  = text[1]
            return KeyValue(key, value, delimitter)
        else:
            return KeyValue(None, None)

    @staticmethod
    def fromAssignOrCharArray(linestr, delimitter='='):
        keyval = linestr.split(delimitter, 1)
        if len(keyval) > 1:
            key    = keyval[0].strip()
            text   = keyval[1].strip().split("'", 2)
            if len(text) >= 2:
                value = text[1]
            else:
                value = text[0]
            return KeyValue(key, value, delimitter)
        else:
            return KeyValue(None, None)

    @staticmethod
    def append(list, key, value):
        if value is not None:
            list.append(key + value)

    def __str__(self):
        return self.keystring + self.value


class KeyValueMap:
    def __init__(self, delimitter='='):
        self.map = {}
        self.delimitter = delimitter

    def readAssign(self, filename):
        with open(filename, 'r') as fp:
            for line in fp:
                kv = KeyValue.fromAssign(line, self.delimitter)
                self.map[kv.key] = kv

    def readAssignString(self, filename):
        with open(filename, 'r') as fp:
            for line in fp:
                kv = KeyValue.fromAssignString(line, self.delimitter)
                self.map[kv.key] = kv

    def readStringArray(self, string_array):
        for part in string_array:
            kv = KeyValue.fromAssign(part, self.delimitter)
            self.map[kv.key] = kv

    def __getitem__(self, key):
        val = self.map.get(key)
        return val.value if val is not None else None

    def __str__(self):
        return str(self.map)

    def __repr__(self):
        return str(self)

File: lua/lib/keyvalue.lua
-- data tools

KeyValue = {}
KeyValue.__index = KeyValue

function KeyValue:new(key, value, delimitter, keystring)
    local obj = {
        key   = key,
        value = value
    }

    if keystring == nil then
        if key == nil then
            obj.keystring = nil
        else
            if delimitter == nil then
                obj.keystring = key .. " "
            else
                obj.keystring = key .. " " .. delimitter .. " "
            end
        end
    else
        obj.keystring = keystring
    end

    setmetatable(obj, KeyValue)
    return obj
end

function KeyValue:valid()
    return self.key ~= nil
end

function KeyValue:has_var()
    return self.value and self.value:find("${", 1, true) ~= nil
end

function KeyValue.from_assign(linestr, delimitter)
    delimitter = delimitter or "="
    local key, value = linestr:match("^(.-)" .. delimitter .. "(.*)$")
    if key then
        key   = key:match("^%s*(.-)%s*$")
        value = value:match("^%s*(.-)%s*$")
        return KeyValue:new(key, value, delimitter)
    end
    return KeyValue:new(nil, nil)
end

function KeyValue.from_assign_string(linestr, delimitter)
    delimitter = delimitter or "="
    local key, rest = linestr:match("^(.-)" .. delimitter .. "(.*)$")
    if key then
        key = key:match("^%s*(.-)%s*$")
        local value = rest:match('"%s*(.-)%s*"')
        return KeyValue:new(key, value, delimitter)
    end
    return KeyValue:new(nil, nil)
end

function KeyValue.from_assign_char_array(linestr, delimitter)
    delimitter = delimitter or "="
    local key, rest = linestr:match("^(.-)" .. delimitter .. "(.*)$")
    if key then
        key = key:match("^%s*(.-)%s*$")
        local value = rest:match("'(.-)'")
        return KeyValue:new(key, value, delimitter)
    end
    return KeyValue:new(nil, nil)
end

function KeyValue.from_assign_or_char_array(linestr, delimitter)
    delimitter = delimitter or "="
    local key, rest = linestr:match("^(.-)" .. delimitter .. "(.*)$")
    if key then
        key = key:match("^%s*(.-)%s*$")
        local value = rest:match("'(.-)'") or rest:match("^%s*(.-)%s*$")
        return KeyValue:new(key, value, delimitter)
    end
    return KeyValue:new(nil, nil)
end

function KeyValue.append(list, key, value)
    if value ~= nil then
        table.insert(list, key .. value)
    end
end

function KeyValue:__tostring()
    return (self.keystring or "") .. (self.value or "")
end



-- KeyValueMap

KeyValueMap = {}
KeyValueMap.__index = KeyValueMap

function KeyValueMap:new(delimitter)
    local obj = {
        map = {},
        delimitter = delimitter or "="
    }
    setmetatable(obj, KeyValueMap)
    return obj
end

function KeyValueMap:read_assign(filename)
    for line in io.lines(filename) do
        local kv = KeyValue.fromAssign(line, self.delimitter)
        self.map[kv.key] = kv
    end
end

function KeyValueMap:read_assign_string(filename)
    for line in io.lines(filename) do
        local kv = KeyValue.fromAssignString(line, self.delimitter)
        self.map[kv.key] = kv
    end
end

function KeyValueMap:read_string_array(arr)
    for _, part in ipairs(arr) do
        local kv = KeyValue.fromAssign(part, self.delimitter)
        self.map[kv.key] = kv
    end
end

function KeyValueMap:get(key)
    local kv = self.map[key]
    return kv and kv.value or nil
end

function KeyValueMap:__tostring()
    local out = "{"
    for k, v in pairs(self.map) do
        out = out .. k .. "=" .. tostring(v) .. ", "
    end
    return out .. "}"
end

